Memory usage optimisations

pjbgf opened the pull request:
The changes improved benchmarks results across number of executions (`+60%`), time (`-4300000 ns/op`), memory (`-23200000 B/op`) and allocations (`-800 allocs/op`).


#### Parse changes Benchmark
```
name       old time/op    new time/op    delta
Parser-16    11.9ms ± 1%     7.4ms ± 1%  -37.77%  (p=0.008 n=5+5)

name       old alloc/op   new alloc/op   delta
Parser-16    30.7MB ± 0%     4.8MB ± 2%  -84.42%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Parser-16     4.19k ± 0%     3.48k ± 0%  -16.84%  (p=0.008 n=5+5)
```

Comparing the memory profile of both executions, shows that the gains are primarily in the zlib writing code.
![image](https://user-images.githubusercontent.com/5452977/198091569-37c19480-2f98-458f-a090-f57742d4bcb1.png)

#### Reference.String() changes Benchmark
For bench mark delta on the ReferenceName optimisations:
```
name                        old time/op    new time/op    delta
ReferenceStringSymbolic-16     140ns ± 4%      40ns ± 9%  -71.19%  (p=0.008 n=5+5)
ReferenceStringHash-16         174ns ±14%      85ns ± 4%  -51.13%  (p=0.008 n=5+5)
ReferenceStringInvalid-16     48.9ns ± 2%     1.5ns ± 3%  -96.96%  (p=0.008 n=5+5)

name                        old alloc/op   new alloc/op   delta
ReferenceStringSymbolic-16     88.0B ± 0%     32.0B ± 0%  -63.64%  (p=0.008 n=5+5)
ReferenceStringHash-16          176B ± 0%      144B ± 0%  -18.18%  (p=0.008 n=5+5)
ReferenceStringInvalid-16      0.00B          0.00B          ~     (all equal)

name                        old allocs/op  new allocs/op  delta
ReferenceStringSymbolic-16      4.00 ± 0%      1.00 ± 0%  -75.00%  (p=0.008 n=5+5)
ReferenceStringHash-16          5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.008 n=5+5)
ReferenceStringInvalid-16       0.00           0.00          ~     (all equal)
```

Comparing source-controller execution without the changes (left) versus with the changes (right), shows that `go_memstats_alloc_bytes_total` seems to be able to release more of the allocated bytes:
![image](https://user-images.githubusercontent.com/5452977/198102346-f0671dc9-6a0f-455e-bb8e-f46d39d8af6c.png)

The setup and configuration are the same as used for testing https://github.com/fluxcd/go-git/pull/4.

pjbgf committed:
Use Sync.Pool pointers to optimise memory usage

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

hiddeco commented on plumbing/reference.go:
```diff
@@ -204,6 +204,20 @@ func (r *Reference) Strings() [2]string {
+		v.WriteString(s)
+	}
+
+	v.WriteString(" ")
```
As `r.Type()` can theoretically have a `InvalidReference` value. Wouldn't it be more accurate to move this in-line of the `case` checks (which already do take the `+ 1` into account)?

pjbgf replied:
@hiddeco Thank you for the feedback. I revisited these changes taking that into account and I think it resulted in a clearer code. The changes did not result in increased allocations nor B/op.

PTAL

hiddeco commented on plumbing/reference_test.go:
```diff
@@ -98,3 +102,11 @@ func (s *ReferenceSuite) TestIsTag(c *C) {
 	c.Assert(r.IsTag(), Equals, true)
 }
+
+func BenchmarkReferenceString(b *testing.B) {
+	r := NewSymbolicReference("v3.1.1", "refs/tags/v3.1.1")
+
+	for n := 0; n < b.N; n++ {
+		r.String()
+	}
+}
```
```suggestion
func benchMarkReferenceString(r *Reference, b *testing.B) {
	for n := 0; n < b.N; n++ {
		r.String()
	}
}

func BenchmarkReferenceString_Symbolic(b *testing.B) {
	benchMarkReferenceString(NewSymbolicReference("v3.1.1", "refs/tags/v3.1.1"), b)
}

func BenchmarkReferenceString_Hash(b *testing.B) {
	benchMarkReferenceString(NewHashReference("v3.1.1", NewHash("6ecf0ef2c2dffb796033e5a02219af86ec6584e5")), b)
}

func BenchmarkReferenceString_Invalid(b *testing.B) {
	benchMarkReferenceString(&Reference{}, b)
}
```

to add varieties.

pjbgf committed:
Optimise Reference.String()
Decreases allocations and bytes per operation by using string builder
with a predefined size.

One additional allocation has been removed by using its own implementation
of Strings(). The reason behind this was due to the fact the calls to
.String() are more recurrent than .Strings() and the performance impact
was worth the code duplication.

Benchmark results:
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
name                        old time/op    new time/op    delta
ReferenceStringSymbolic-16     140ns ± 4%      40ns ± 9%  -71.19%  (p=0.008 n=5+5)
ReferenceStringHash-16         174ns ±14%      85ns ± 4%  -51.13%  (p=0.008 n=5+5)
ReferenceStringInvalid-16     48.9ns ± 2%     1.5ns ± 3%  -96.96%  (p=0.008 n=5+5)

name                        old alloc/op   new alloc/op   delta
ReferenceStringSymbolic-16     88.0B ± 0%     32.0B ± 0%  -63.64%  (p=0.008 n=5+5)
ReferenceStringHash-16          176B ± 0%      144B ± 0%  -18.18%  (p=0.008 n=5+5)
ReferenceStringInvalid-16      0.00B          0.00B          ~     (all equal)

name                        old allocs/op  new allocs/op  delta
ReferenceStringSymbolic-16      4.00 ± 0%      1.00 ± 0%  -75.00%  (p=0.008 n=5+5)
ReferenceStringHash-16          5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.008 n=5+5)
ReferenceStringInvalid-16       0.00           0.00          ~     (all equal)

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

hiddeco reviewed (approved):
Thanks a lot @pjbgf :bow:
